pkg/providers/vsphere/vsphere.go

```diff
@@ -245,109 +225,25 @@ func (p *vsphereProvider) MachineResourceType() string {
 	return eksaVSphereMachineResourceType
 }
 
-func (p *vsphereProvider) setupSSHAuthKeysForCreate() error {
-	var useKeyGeneratedForControlplane, useKeyGeneratedForWorker bool
-	var err error
-	controlPlaneUser := p.machineConfigs[p.clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users[0]
-	p.controlPlaneSshAuthKey = controlPlaneUser.SshAuthorizedKeys[0]
-	if len(p.controlPlaneSshAuthKey) > 0 {
-		p.controlPlaneSshAuthKey, err = common.StripSshAuthorizedKeyComment(p.controlPlaneSshAuthKey)
-		if err != nil {
-			return err
-		}
-	} else {
-		logger.Info("Provided control plane sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-		generatedKey, err := common.GenerateSSHAuthKey(p.writer)
-		if err != nil {
-			return err
-		}
-		p.controlPlaneSshAuthKey = generatedKey
-		useKeyGeneratedForControlplane = true
-	}
-	controlPlaneUser.SshAuthorizedKeys[0] = p.controlPlaneSshAuthKey
-	for _, workerNodeGroupConfiguration := range p.clusterConfig.Spec.WorkerNodeGroupConfigurations {
-		workerUser := p.machineConfigs[workerNodeGroupConfiguration.MachineGroupRef.Name].Spec.Users[0]
-		p.workerSshAuthKey = workerUser.SshAuthorizedKeys[0]
-		if len(p.workerSshAuthKey) > 0 {
-			p.workerSshAuthKey, err = common.StripSshAuthorizedKeyComment(p.workerSshAuthKey)
-			if err != nil {
-				return err
-			}
-		} else {
-			if useKeyGeneratedForControlplane { // use the same key
-				p.workerSshAuthKey = p.controlPlaneSshAuthKey
-			} else {
-				logger.Info("Provided worker sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-				generatedKey, err := common.GenerateSSHAuthKey(p.writer)
-				if err != nil {
-					return err
-				}
-				p.workerSshAuthKey = generatedKey
-				useKeyGeneratedForWorker = true
-			}
-		}
-		workerUser.SshAuthorizedKeys[0] = p.workerSshAuthKey
-	}
-	if p.clusterConfig.Spec.ExternalEtcdConfiguration != nil {
-		etcdUser := p.machineConfigs[p.clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users[0]
-		p.etcdSshAuthKey = etcdUser.SshAuthorizedKeys[0]
-		if len(p.etcdSshAuthKey) > 0 {
-			p.etcdSshAuthKey, err = common.StripSshAuthorizedKeyComment(p.etcdSshAuthKey)
-			if err != nil {
-				return err
-			}
-		} else {
-			if useKeyGeneratedForControlplane { // use the same key as for controlplane
-				p.etcdSshAuthKey = p.controlPlaneSshAuthKey
-			} else if useKeyGeneratedForWorker {
-				p.etcdSshAuthKey = p.workerSshAuthKey // if cp key was provided by user, check if worker key was generated by cli and use that
+func (p *vsphereProvider) generateSSHKeysIfNotSet() error {
+	var generatedKey string
+	for _, machineConfig := range p.machineConfigs {
+		user := machineConfig.Spec.Users[0]
+		if user.SshAuthorizedKeys[0] == "" {
+			if generatedKey != "" { // use the same key
+				user.SshAuthorizedKeys[0] = generatedKey
```

Comment:
`generatedKey` is always gonna be empty at this point right?

Reply:
No necessarily. It can have been generated in prior iterations of the loop. The first machine config that is found without an ssh key will encounter an empty `generatedKey` which will trigger an SSH key generation. The next machine config (if there is any) without a key, will reuse `generatedKey`

Reply:
oh totally missed the for loop